### PR TITLE
obs-outputs: Add eRTMP/eFLV multitrack support

### PR DESCRIPTION
**Pull request by palana**
### Description
Add eRTMP multitrack video support for the rtmp output and "full" eFLV support for the FLV output (both additional video codecs and multitrack video)

### Motivation and Context
This allows transmitting multiple video streams within the same RTMP connection; it's used for Twitch Enhanced Broadcasting to allow multiple renditions to be sent through a single RTMP connection, helping ensure that renditions remain aligned

### How Has This Been Tested?
It's being used in the Twitch Enhanced Broadcasting Beta

### Types of changes
- New feature (non-breaking change which adds functionality)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Comment by RytoEX** [comment]
With https://github.com/obsproject/obs-studio/pull/10494 merged, this can be rebased and undrafted.

**Comment by palana** [comment]
> With #10494 merged, this can be rebased and undrafted.

@RytoEX: done

File: plugins/obs-outputs/flv-mux.c [hunk 1]
```diff
@@ -371,30 +389,46 @@ void flv_packet_ex(struct encoder_packet *packet, enum video_id_t codec_id,
...
+	if (is_multitrack) {
+		s_w8(&s, MULTITRACKTYPE_ONE_TRACK | type);
+		s_w4cc(&s, codec_id);
+		// trackId
+		s_w8(&s, (uint8_t)idx);
+	} else {
+		s_w4cc(&s, codec_id);
+	}
+
+	// h264/hevc composition time offset
```
Comment (RytoEX) [inline plugins/obs-outputs/flv-mux.c:427]:
```suggestion
	// H.264/HEVC composition time offset
```

File: plugins/obs-outputs/flv-output.c [hunk 2]
```diff
@@ -41,12 +46,101 @@ struct flv_output {
...
+/* Adapted from FFmpeg's libavutil/pixfmt.h
+ *
+ * Renamed to make it apparent that these are not imported as this module does
+ * not use or link against FFmpeg.
+ */
+
+/* clang-format off */
+
+/**
+  * Chromaticity coordinates of the source primaries.
```
Comment (RytoEX) [inline plugins/obs-outputs/flv-output.c:66]:
```suggestion
  * Chromaticity coordinates of the source primaries
```
or
```suggestion
  * Define chromaticity coordinates of the source primaries
```

The first is more "label-like" or "header-like" and would be appropriate here.

File: plugins/obs-outputs/flv-output.c [hunk 3]
```diff
@@ -41,12 +46,101 @@ struct flv_output {
...
+	OBSCOL_PRI_SMPTEST428_1 = OBSCOL_PRI_SMPTE428,
+	OBSCOL_PRI_SMPTE431     = 11, ///< SMPTE ST 431-2 (2011) / DCI P3
+	OBSCOL_PRI_SMPTE432     = 12, ///< SMPTE ST 432-1 (2010) / P3 D65 / Display P3
+	OBSCOL_PRI_EBU3213      = 22, ///< EBU Tech. 3213-E (nothing there) / one of JEDEC P22 group phosphors
+	OBSCOL_PRI_JEDEC_P22    = OBSCOL_PRI_EBU3213,
+	OBSCOL_PRI_NB                 ///< Not part of ABI
+};
+
+/**
+ * Color Transfer Characteristic.
```
Comment (RytoEX) [inline plugins/obs-outputs/flv-output.c:90]:
```suggestion
 * Color Transfer Characteristic
```

File: plugins/obs-outputs/flv-output.c [hunk 4]
```diff
@@ -41,12 +46,101 @@ struct flv_output {
...
+	OBSCOL_TRC_SMPTE2084    = 16, ///< SMPTE ST 2084 for 10-, 12-, 14- and 16-bit systems
+	OBSCOL_TRC_SMPTEST2084  = OBSCOL_TRC_SMPTE2084,
+	OBSCOL_TRC_SMPTE428     = 17, ///< SMPTE ST 428-1
+	OBSCOL_TRC_SMPTEST428_1 = OBSCOL_TRC_SMPTE428,
+	OBSCOL_TRC_ARIB_STD_B67 = 18, ///< ARIB STD-B67, known as "Hybrid log-gamma"
+	OBSCOL_TRC_NB                 ///< Not part of ABI
+};
+
+/**
+ * YUV colorspace type.
```
Comment (RytoEX) [inline plugins/obs-outputs/flv-output.c:119]:
```suggestion
 * YUV colorspace type
```

File: plugins/obs-outputs/flv-output.c [hunk 5]
```diff
@@ -111,42 +236,211 @@ static void write_meta_data(struct flv_output *stream)
...
+	switch (stream->video_codec[idx]) {
+	case CODEC_NONE:
+		do_log(LOG_ERROR,
+		       "Codec not initialized for track %zu while sending header",
+		       idx);
+		return false;
+
+	case CODEC_H264:
+		packet.size = obs_parse_avc_header(&packet.data, header, size);
+		// Always send H264 on track 0 as old style for compat
```
Comment (RytoEX) [inline plugins/obs-outputs/flv-output.c:282]:
```suggestion
		// Always send H.264 on track 0 as old style for compatibility.
```

**Review by RytoEX — CHANGES_REQUESTED** [review]
Nits.

File: plugins/obs-outputs/flv-mux.c [hunk 6]
```diff
@@ -59,22 +66,32 @@ enum datatype_t {
 static void s_w4cc(struct serializer *s, enum video_id_t id)
 {
 	switch (id) {
+	case CODEC_NONE:
+		assert(0 && "Tried to serialize CODEC_NONE");
```
Comment (PatTheMav) [inline plugins/obs-outputs/flv-mux.c:70]: We set `NDEBUG` for all configurations apart from `Debug`, so the assert is inert in those. Is that taken into account here?

Comment (RytoEX) [reply plugins/obs-outputs/flv-mux.c:70]: I briefly checked with @notr1ch off-thread and it seems that assert here is okay. @palana could you summarize the intent with the assert here to make sure we're all on the same page?

**Review by Lain-B — APPROVED** [review]
Outside of the ryan/pat nitpicks this has my approval. An alternative for the assert might be like, bcrash() or something. (If it's not just a debug situation at least)

**Comment by palana** [comment]
@RytoEX: nits should be addressed and squashed (and rebased)

File: plugins/obs-outputs/rtmp-stream.c [hunk 7]
```diff
@@ -922,33 +961,41 @@ static inline bool send_headers(struct rtmp_stream *stream)
...
+	for (size_t i = 0; i < MAX_OUTPUT_VIDEO_ENCODERS; i++) {
+		obs_encoder_t *encoder =
+			obs_output_get_video_encoder2(stream->output, i);
+		if (!encoder)
+			return true;
 
-	// Y2023 spec
-	return send_video_footer(stream);
+		if (stream->video_codec[i] == CODEC_H264)
+			continue;
```
Comment (derrod) [inline plugins/obs-outputs/rtmp-stream.c:992]: If H.264 is not on track 0, i.e. sent as "enhanced" packet, it should also send a footer.

File: plugins/obs-outputs/rtmp-stream.c [hunk 8]
```diff
@@ -922,33 +961,41 @@ static inline bool send_headers(struct rtmp_stream *stream)
...
 	}
 
+	i = 1;
+	while (obs_output_get_video_encoder2(stream->output, i) != NULL) {
+		if (!send_video_metadata(stream, i) ||
+		    !send_video_header(stream, i))
+			return false;
+		i += 1;
+	}
+
```
Comment (derrod) [inline plugins/obs-outputs/rtmp-stream.c:979]: Is there any situation where there could be a gap in encoders? i.e. 0, 1, 3 are valid but 2 is empty?

Either way I'd personally write this as a for-loop:
```c
for (size_t idx = 1; idx < MAX_OUTPUT_VIDEO_ENCODERS; idx++) {
    obs_encoder_t *enc = obs_output_get_video_encoder2(stream->output, idx);
    if (!enc)
        continue;
    if (!send_video_metadata(stream, idx) || !send_video_header(stream, idx))
        return false;
}
```

File: plugins/obs-outputs/rtmp-stream.c [hunk 9]
```diff
@@ -922,33 +961,41 @@ static inline bool send_headers(struct rtmp_stream *stream)
...
-	// send metadata only if HDR
-	video_t *video = obs_get_video();
-	const struct video_output_info *info = video_output_get_info(video);
-	enum video_colorspace colorspace = info->colorspace;
-	if (colorspace == VIDEO_CS_2100_PQ || colorspace == VIDEO_CS_2100_HLG)
-		if (!send_video_metadata(stream)) // Y2023 spec
-			return false;
-
-	if (!send_video_header(stream))
+	if (!send_video_metadata(stream, 0) || !send_video_header(stream, 0))
```
Comment (derrod) [inline plugins/obs-outputs/rtmp-stream.c:964]: Also in general, is there a reason this is split from the metadata for idx > 0? Why not send all of them in one big loop?

**Review by derrod — CHANGES_REQUESTED** [review]
A couple smaller things, it would be good if you could also adjust `init_send()` to correctly accumulate bitrate from all video encoders rather than just the primary one, that should fix the issues peopole have seen with the network enhancement options, but that's not a blocker for now.

File: plugins/obs-outputs/flv-mux.c (hunk 6 above)
Comment (RytoEX) [reply plugins/obs-outputs/flv-mux.c:70]: This is fine as-is for now then. If need be, we'll re-evaluate later.

**Review by RytoEX — CHANGES_REQUESTED** [review]
Nits. I'll probably pass after this round.

File: plugins/obs-outputs/rtmp-stream.c (hunk 8 above)
Comment (palana) [reply plugins/obs-outputs/rtmp-stream.c:979]: the beta multitrack video output doesn't create gaps, but there shouldn't be a reason to not support them

Comment (palana) [reply plugins/obs-outputs/rtmp-stream.c:979]: @derrod: might be interesting to also change this for audio tracks

Comment (palana) [reply plugins/obs-outputs/rtmp-stream.c]: changed in https://github.com/obsproject/obs-studio/pull/10500#event-12641499492

File: plugins/obs-outputs/rtmp-stream.c (hunk 9 above)
Comment (palana) [reply plugins/obs-outputs/rtmp-stream.c:964]: not as far as I'm aware, no; @kc5nra: do you know if there's a reason to split headers/metadata like this?

File: plugins/obs-outputs/rtmp-stream.c (hunk 7 above)
Comment (palana) [reply plugins/obs-outputs/rtmp-stream.c]: changed in https://github.com/obsproject/obs-studio/pull/10500#commits-pushed-bac01a2

File: plugins/obs-outputs/rtmp-stream.c (hunk 9 above)
Comment (RytoEX) [reply plugins/obs-outputs/rtmp-stream.c:964]: This feedback thread is still active.

File: plugins/obs-outputs/flv-mux.c [hunk 10]
```diff
@@ -371,30 +389,46 @@ void flv_packet_ex(struct encoder_packet *packet, enum video_id_t codec_id,
...
 	s_wtimestamp(&s, time_ms);
 	s_wb24(&s, 0); // always 0
 
 	// packet ext header
-	s_w8(&s,
-	     FRAME_HEADER_EX | type | (packet->keyframe ? FT_KEY : FT_INTER));
-	s_w4cc(&s, codec_id);
+	s_w8(&s, FRAME_HEADER_EX |
+			 (is_multitrack ? PACKETTYPE_MULTITRACK : type) |
+			 (packet->keyframe ? FT_KEY : FT_INTER));
```
[the remark this answers is not in the document]
Comment (RytoEX) [reply plugins/obs-outputs/flv-mux.c:412]: This feedback thread is still active.

File: plugins/obs-outputs/rtmp-stream.c (hunk 9 above)
Comment (palana) [reply plugins/obs-outputs/rtmp-stream.c]: changed in https://github.com/obsproject/obs-studio/pull/10500#commits-pushed-cf791ae

File: plugins/obs-outputs/flv-mux.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (palana) [reply plugins/obs-outputs/flv-mux.c]: changed in https://github.com/obsproject/obs-studio/pull/10500#commits-pushed-cf791ae

**Comment by RytoEX** [comment]
Off-thread discussion indicates that this looks good. Squash the commits appropriately and we'll give it a (hopefully) final pass.

**Comment by palana** [comment]
> Off-thread discussion indicates that this looks good. Squash the commits appropriately and we'll give it a (hopefully) final pass.

@RytoEX: squashed and rebased

**Review by RytoEX — APPROVED** [review]
